### ISSUE TITLE
swapped x-y field components when importing eigenvector

### ISSUE THREAD

**Issue by M8kmyday** [mfem/mfem#2637]
I am writing a 2D electromagnetic simulator with losses.  I set it up in MFEM then solve it in a separate C program using PETSc and SLEPc compiled for complex.  Bringing the eigenvector back into MFEM, then plotting, results in the x- and y- electric field components being swapped.  I don't know if the eigenvector needs to be re-ordered before importing, if the eigenvector is fine but somewhere inside MFEM the x and y components are being swapped, or something else.  So I am looking for feedback on how to resolve the problem.  A detailed explanation of what I am doing is attached.

[WR90_results.pdf](https://github.com/mfem/mfem/files/7470944/WR90_results.pdf)

**Comment by M8kmyday** [comment]
I am seeing the same problem with ex13p.  The example only shows magnitudes in glvis, so I added ParaView output to get a vector plot.  The only modification I made is to insert the following code before line 220:

         // ParaView
         ParaViewDataCollection *pd=NULL;

         pd=new ParaViewDataCollection(mode_name.str().c_str(),pmesh);
         pd->SetPrefixPath("ParaView");
         pd->RegisterField("eigenmode",&x);
         pd->SetLevelsOfDetail(order);
         pd->SetDataFormat(VTKFormat::ASCII);
         pd->SetHighOrderOutput(true);
         pd->Save();
and I call my version ex13pby.cpp.  The makefile is modified to add ex13pby, then it is compiled with the rest of the parallel examples.

I then run the example as 
mpirun -np 1  ex13pby -m ../data/square-disc.mesh -o 2 -n 4

Looking at mode 0, the magnitude of the fields look correct and similar to the results shown by glvis:

![Screenshot from 2021-11-09 11-02-29](https://user-images.githubusercontent.com/90462987/140970353-4ce70adc-440c-4d52-8547-f698da3161b2.png)

Adding a glyph filter to get a vector, 

![Screenshot from 2021-11-09 11-05-39](https://user-images.githubusercontent.com/90462987/140970979-714d3382-54dd-435e-8bcf-f6bca861724e.png)

I see only x-directed vectors, the same result I get from my work.  The vector violates Maxwell's equations since the vector should point orthogonally to the central hole.

Plotting the x-component only, I get:

![Screenshot from 2021-11-09 11-08-31](https://user-images.githubusercontent.com/90462987/140971341-ac9557bd-1656-4eab-9c21-6daa1aa6aab7.png)

This looks correct.

Plotting the y-component, I get:

![Screenshot from 2021-11-09 11-09-59](https://user-images.githubusercontent.com/90462987/140971551-56ca55cd-c84a-45e8-bfc5-986fcccc6101.png)

The vectors do not turn to the y-direction, but if they did, this plot would be correct.

Can you check this example on your end and see if you see what I see?

**Comment by pazner** [comment]
Try converting the 2D vector field to a 3D vector field using the "Calculator" filter before plotting the glyphs. I think ParaView's Glyph filter doesn't handle the 2D vector fields well.

This is what I get:

<img width="1105" alt="image" src="https://user-images.githubusercontent.com/11493037/140981605-cad9597f-a28e-41fc-88e3-93f385c3abab.png">

**Comment by M8kmyday** [comment]
That works for me, too.  Thanks!

I can see in the ParaView text files for my work that the y-component is 6 to 7 orders of magnitude smaller than the x-component, so the x-component really does dominate when it should be the y-component.  So when I convert to a vector with the Calculator filter, the plot remains unchanged.

Now seeing ex13p plotting correctly, I can run that example in my code and look for differences.  The two cases are pretty similar but with my implementation including the z-component of the electric field to handle losses plus the built-in assumption of propagation.  Still, I may be able to make some headway with this approach.

For what it is worth, I have used gdb to verify that no function with "RT" in the name is called in my implementation.  Everything is strictly "ND".  This goes back to an observation via email from Mark that maybe RT elements were being pulled in somewhere.

**Comment by M8kmyday** [comment]
I do not have the boundary conditions set correctly.  I can make ex13p give the same incorrect field structure as my code by commenting out the boundary conditions (i.e. the EliminateEssentialBCDiag calls).  My setup is more complex than ex13p, so I'm working on how to set the boundary conditions.

**Comment by M8kmyday** [comment]
Ok, I finally got it.  Here is the correct field for the rectangular waveguide:
![Screenshot from 2021-11-13 09-54-09](https://user-images.githubusercontent.com/90462987/141650358-680590e8-bdbc-42f9-a1c3-0cf74135f23b.png)

I ended up exporting the essential true dofs, importing them to my C program, then applying the boundary conditions there by zeroing out rows/columns with a constant on the diagonal using PETSc function calls.

The zeroing/constant operation within MFEM seems to be limited to a single matrix.  My basic formulation looks like this:
![Screenshot from 2021-11-13 09-56-46](https://user-images.githubusercontent.com/90462987/141650491-3b44f9bb-79a3-4020-b893-73ef54c63b84.png)
so I have multiple matrices, and matrices sum to make the field.  I need to apply the zeroing/constant operation to the matrix sum, but I could not figure out how to get the sum done within MFEM.

At some point, I would like to apply the boundary conditions within MFEM, then just solve the generalized and complex-valued eigenvalue problem in the C program.

One more example:
![Screenshot from 2021-11-13 10-28-36](https://user-images.githubusercontent.com/90462987/141651372-d7a46f30-4913-4997-8109-0d9918e90097.png)
